File: Sep/Day16/replace-non-coprime-numbers-in-array.java
import java.util.*;

class Solution {
    public List<Integer> replaceNonCoprimes(int[] nums) {
        List<Integer> stack = new ArrayList<>();
        for (int num : nums) {
            while (!stack.isEmpty()) {
                int g = gcd(stack.get(stack.size() - 1), num);
                if (g == 1) break;
                num = (int)((1L * stack.remove(stack.size() - 1) * num) / g); // cast to long to avoid overflow
            }
            stack.add(num);
        }
        return stack;
    }

    private int gcd(int a, int b) {
        while (b != 0) {
            int temp = b;
            b = a % b;
            a = temp;
        }
        return a;
    }
}

File: Sep/Day16/replace-non-coprime-numbers-in-array.cc
#include <vector>
#include <numeric> // for gcd
using namespace std;

class Solution {
public:
    vector<int> replaceNonCoprimes(vector<int>& nums) {
        vector<int> stack;
        for (int num : nums) {
            while (!stack.empty()) {
                int g = gcd(stack.back(), num);
                if (g == 1) break;
                num = (1LL * stack.back() * num) / g; // use long long to prevent overflow
                stack.pop_back();
            }
            stack.push_back(num);
        }
        return stack;
    }
};
